substratevm/src/com.oracle.svm.hosted/src/com/oracle/svm/hosted/image/NativeImageBFDNameProvider.java

```diff
@@ -383,14 +420,163 @@ public void setNativeLibs(NativeLibraries nativeLibs) {
+         * ==> Hello* i.e. both S_ and S1_ would demangle to Hello. This situation should never
+         * arise. A name can always be correctly encoded without repeats. In the above example that
+         * woudl be _ZN5Hello9compareToEJiPS_.
+         */
         final NativeImageBFDNameProvider nameProvider;
         final StringBuilder sb;
-        final List<String> prefixes;
+
+        // A list of lookup names identifying substituted elements. A prospective name for an
+        // element that is about to be encoded can be looked up in this list. If a match is found
+        // the list index can be used to identify the relevant short symbol. If it is not found
+        // then inserting the name serves to allocate the short name associated with the inserted
+        // element's result index.
+        List<LookupName> bindings;
+
+        /**
+         * A lookup name is used as a key to record and subsequently lookup a short symbol that can
+         * replace one or more elements of a mangled name. A lookup name is usually just a simple
+         * string, i.e. some text that is to be mangled and to which the corresponding short symbol
+         * should decode. However, substitutions for namespace prefixes (e.g. AppLoader506::Foo)
+         * require a composite lookup name that composes a namespace prefix (AppLoader506) with a
+         * trailing simple name (Foo). The corresponding short symbol will demangle to the string
+         * produced by composing the prefix and simple name with a :: separator (i.e. restoring
+         * AppLoader506::Foo). Short symbols for pointer types (e.g. Foo* or AppLoader506::Foo*)
+         * require a pointer lookup name that identifies the substituted text as a pointer to some
+         * underlying type (e.g. Foo or AppLoader506::Foo). The corresponding short symbol will
+         * demangle to the string produced by demangling the underlying type with a * suffix. In
+         * theory the prefix for a pointer type lookup name might be defined by another pointer type
+         * lookup name (if, say, we needed to encode type Foo**). In practice, that case should not
+         * arise with Java method signatures.
+         */
+        private abstract class LookupName {
+            String value;
+
+            protected LookupName(String value) {
+                assert value != null;
+                this.value = value;
+            }
+
+            @Override
+            public abstract String toString();
+        }
+
+        private class SimpleLookupName extends LookupName {
+            SimpleLookupName(String value) {
+                super(value);
+            }
+
+            @Override
+            public boolean equals(Object other) {
+                if (other instanceof SimpleLookupName) {
+                    return this.value.equals(((SimpleLookupName) other).value);
+                }
```

Comment:
TIL (so I thought you might want to learn some ["new", since Java 14, tricks](https://openjdk.org/jeps/305) as well) you can do 
```suggestion
                if (other instanceof SimpleLookupName simpleLookupName) {
                    return this.value.equals(simpleLookupName.value);
                }
```

Reply:
Oh my, I am an old dog and find these new tricks so very challenging . . .

Reply:
If it makes you feel any better I find it hard to keep track of these as well  (especially remembering in which version each trick works) despite being a younger dog :)